linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1713,6 +1716,268 @@ private[optimizer] abstract class OptimizerCore(
+    import EvalContextInsertion._
+
+    object valTreeInfo extends Traversers.Traverser {
+      val mutatedLocalVars = mutable.Set.empty[LocalName]
+
+      traverse(valTree)
+
+      override def traverse(tree: Tree): Unit = {
+        super.traverse(tree)
+        tree match {
+          case Assign(VarRef(ident), _) => mutatedLocalVars += ident.name
+          case _                        => ()
+        }
+      }
+    }
+
+    def recs(bodies: List[Tree]): EvalContextInsertion[List[Tree]] = bodies match {
+      case Nil =>
+        NotFoundPureSoFar
+      case firstBody :: restBodies =>
+        rec(firstBody) match {
+          case Success(newFirstBody) => Success(newFirstBody :: restBodies)
+          case NotFoundPureSoFar     => recs(restBodies).mapOrKeepGoing(firstBody :: _)
+          case Failed                => Failed
+        }
+    }
+
+    def rec(body: Tree): EvalContextInsertion[Tree] = {
+      implicit val pos = body.pos
+
+      body match {
+        case VarRef(ident) =>
+          if (ident.name == valName)
+            Success(valTree)
+          else if (valTreeInfo.mutatedLocalVars.contains(ident.name))
+            Failed
+          else
+            NotFoundPureSoFar
+
+        case Skip() =>
+          NotFoundPureSoFar
+
+        case Block(stats) =>
+          recs(stats).mapOrKeepGoing(Block(_))
+
+        case Labeled(label, tpe, innerBody) =>
+          rec(innerBody).mapOrKeepGoing(Labeled(label, tpe, _))
+
+        case Return(expr, label) =>
+          rec(expr).mapOrFailed(Return(_, label))
+
+        case If(cond, thenp, elsep) =>
+          rec(cond).mapOrFailed(If(_, thenp, elsep)(body.tpe))
+
+        case Throw(expr) =>
+          rec(expr).mapOrFailed(Throw(_))
+
+        case Match(selector, cases, default) =>
+          rec(selector).mapOrFailed(Match(_, cases, default)(body.tpe))
+
+        case New(className, ctor, args) =>
+          recs(args).mapOrKeepGoingIf(New(className, ctor, _))(
+              keepGoingIf = hasElidableConstructors(className))
+
+        case LoadModule(className) =>
+          if (hasElidableConstructors(className)) NotFoundPureSoFar
+          else Failed
+
+        case Select(qual, field) =>
+          rec(qual).mapOrFailed(Select(_, field)(body.tpe))
+
+        case Apply(flags, receiver, method, args) =>
+          rec(receiver) match {
+            case Success(newReceiver) =>
+              Success(Apply(flags, newReceiver, method, args)(body.tpe))
+            case NotFoundPureSoFar if isNotNull(receiver) =>
+              recs(args).mapOrFailed(Apply(flags, receiver, method, _)(body.tpe))
+            case _ =>
+              Failed
+          }
+
+        case ApplyStatically(flags, receiver, className, method, args) =>
+          rec(receiver) match {
+            case Success(newReceiver) =>
+              Success(ApplyStatically(flags, newReceiver, className, method, args)(body.tpe))
+            case NotFoundPureSoFar if isNotNull(receiver) =>
```

Comment:
`if nullPointersAreUnchecked || isNotNull(receiver)`?

Reply:
Ah ah, I was about to explain why we cannot do it here, and then I realized the explanation applies to the other places where I looked at `Unchecked` behaviors. So in fact we can never do that. I expanded the doc comment on the method to explain why we cannot do it.